### ISSUE TITLE
Non-callable @helpers

### ISSUE THREAD

**Issue by rpelizzi** [zappajs/zappajs#81]
According to the docs, there are two kind of helpers: non-callable and callable. The former is evaluated at app startup, so it cannot depend on stuff defined in the request handler scope. The latter is evaluated at call time, e.g. `@myhelper()` and has access to the request scope.

There is a third scenario that doesn't seem to be covered: non-callable values that depend on the request scope. Example: define a helper `@user` such that  `@user == @req.user`. Currently, you can only define `@user` to be a callable that returns `@req.user`.

I can't even write a middleware that does that, can I? AFAIK Zappa does not let you hook into its "scope filling" process.

**Comment by shimaore** [comment]
`@` in Zappa requests and middleware is volatile. But Express provides you with the tools to do the job; you can store the value as either `@session.user` (would persists across a session) or `@locals.user` (bound to a single HTTP request) in a scenario like this.

You can do this in ZappaJS middleware but there's a catch (which in my mind looks a lot like a missing feature in ZappaJS): you can only use ZappaJS middleware "inline", just like so:

```
require('zappajs') ->
  # Pick one of mw1|mw2
  mw1 = ->
    @session.user = @req.user
    @next()

  mw2 = ->
    @locals.user = @req.user
    @next()

  @get '/1', mw1, -> # do something
```

I've never understood why we couldn't do this with `@use`, and looking at the code, it won't take much to get us to:

```
require('zappajs') ->  # not working yet
  @use ->
    @session.user = @req.user
    @next()
```

Maybe the reason for not doing so was to avoid performance penalty. In that case we could build an idiom:

```
require('zappajs') ->  # not working yet
  @use @middleware ->
    @session.user = @req.user
    @next()
```

where @middleware wraps a Zappa function into a middleware handle. We could keep backward-compatibility with a setting (say, `magic-middleware`) which would default to `on` and make `@middleware` a no-op when `on`. Turn it `off` and only middleware you tagged as ZappaJS middleware gets wrapped.

Anyone got feedback before I get coding on this?

**Comment by rpelizzi** [comment]
Of course I could bind it to the `@session` or `@req` objects (it's already in my scope as `@req.user`). This is purely about creating a shorthand for convenience. I just thought it made sense we could define our own shortcuts on the context: if zappa provides `@send` such that `@send == @res.send`, why not allow users to define their own `@myVar` such that `@myVar == @req.myVar`? Once you create these shorthands into a request variable, you can easily attach them to every context as you wrap it for zappa middleware functions and views.

BTW with the same logic, we could have an `@error` "middleware-marker" so that you can wrap error middleware as well:

```
@use @error ->
  console.log @error
  @send @error.stack, 500
```

I had to define an error handler and I could not find a way to avoid using the 4-arg notation. Did I miss something?

**Comment by shimaore** [comment]
What you want looks like [(Lisp-ish) macros](http://blog.davidpadbury.com/2010/12/09/making-macros-in-coffeescript/).

This could also be done in ZappaJS without macros but we'd have to set / retrieve the variables before / after every middleware and handler callbacks. Would have to test to know how bad that would be performance-wise.

Looking at [proto.js](https://github.com/senchalabs/connect/blob/master/lib/proto.js) I'm not sure how `@error` would hook up with that code.

**Comment by rpelizzi** [comment]
yeah that's a lot of work for a shorthand. it could be interesting, but only if it were easier to implement.

I can't help you much with `@error`... I would think that whatever we do to wrap the handlers we define for `@use` (which take no argument) into a function with arity 3 that then maps the arguments into the `@req`, `@res` and `@next` and applies the context to the zappa middleware can be done separately, just by wrapping the 0-arity handler into a 4-arity handler instead of 3-arity. However I can't seem to find the 3-arity step in the code...

**Comment by shimaore** [comment]
FWIW the `@middleware` pattern was added in 79d1dd2d6323fa92f22d84b0d6655f7148279a73, which also introduced

```
@enable 'magic middleware'
```

This is in release 0.4.22.

> I had to define an error handler and I could not find a way to avoid using the 4-arg notation. Did I miss something?

Re-reading the code for proto.js, this seems to be the only way to catch those. It'd be nice to have an example in the ZappaJS doc though.

**Comment by shimaore** [comment]
Note to self: there is an attempt at writing a `@error` handle in the `handle_errors` branch, however it appears I didn't get it right. Work in progress.
